libdnf/base/goal.cpp

```diff
@@ -1690,6 +1693,97 @@ void Goal::Impl::add_paths_to_goal() {
     rpm_filepaths.clear();
 }
 
+void Goal::Impl::set_exclude_from_weak(const std::vector<std::string> & exclude_from_weak) {
+    for (const auto & exclude_weak : exclude_from_weak) {
+        rpm::PackageQuery weak_query(base, rpm::PackageQuery::ExcludeFlags::APPLY_EXCLUDES);
+        libdnf::ResolveSpecSettings settings{
+            .with_nevra = true, .with_provides = false, .with_filenames = true, .with_binaries = false};
+        weak_query.resolve_pkg_spec(exclude_weak, settings, false);
+        weak_query.filter_available();
+        rpm_goal.add_exclude_from_weak(*weak_query.p_impl);
+    }
+}
+
+void Goal::Impl::autodetect_unsatisfied_installed_weak_dependencies() {
+    rpm::PackageQuery installed_query(base, rpm::PackageQuery::ExcludeFlags::IGNORE_EXCLUDES);
+    installed_query.filter_installed();
+    if (installed_query.empty()) {
+        return;
+    }
+    rpm::PackageQuery base_query(base, rpm::PackageQuery::ExcludeFlags::APPLY_EXCLUDES);
+    rpm::ReldepList reldep_list(base);
+
+    std::vector<std::string> installed_names;
+    installed_names.reserve(installed_query.size());
+
+    // Investigate uninstalled recommends of installed packages
+    for (auto pkg : installed_query) {
+        installed_names.push_back(pkg.get_name());
+        for (auto recommend : pkg.get_recommends()) {
```

Comment:
There's probably no reason not having the `ReldepList::add(const Reldep &);`. After making it const we could switch also here and below in supplements to iterate using `const auto &`, but it's not anything major I guess...